### PR TITLE
use merge key list instead of a single merge key

### PR DESCRIPTION
**Pull request by natasha41575**
ref: #3111 

Still a draft, but the test I added to patchstrategicmerge_test.go passes

Working on edge cases (will add tests) and cleaning up & removing some redundancy in the code

### CODE REVIEW

File: kyaml/yaml/fns.go [hunk 1]
```diff
@@ -68,7 +72,7 @@ func (e ElementSetter) isMappingSetter() bool {
 	return e.Key != "" && e.Value != ""
 }
```
Comment (yuwenma) [inline kyaml/yaml/fns.go:74]: Filter is an exportable function. Is that okay to change the function name? Will it break the some potential callers?

Comment (yuwenma) [reply kyaml/yaml/fns.go:74]: oh, I see Filter not removed.

File: kyaml/yaml/fns.go [hunk 2]
```diff
@@ -68,7 +72,7 @@ func (e ElementSetter) isMappingSetter() bool {
...
 
-func (e ElementSetter) Filter(rn *RNode) (*RNode, error) {
+func (e ElementSetter) FilterPatch(rn *RNode) (*RNode, error) {
```
Comment (yuwenma) [inline kyaml/yaml/fns.go:75]: nit: suggest adding docstring for exportable functions.

File: kyaml/yaml/rnode.go [hunk 3]
```diff
@@ -474,13 +474,32 @@ func (rn *RNode) ElementValues(key string) ([]string, error) {
 	return elements, nil
 }
 
+func (rn *RNode) ElementValuesList(keys []string) ([][]string, error) {
```
Comment (yuwenma) [inline kyaml/yaml/rnode.go:477]: ditto: docstring

File: kyaml/yaml/fns.go [hunk 4]
```diff
@@ -205,8 +272,8 @@ func (c FieldClearer) Filter(rn *RNode) (*RNode, error) {
 	return nil, nil
 }
 
-func MatchElement(field, value string) ElementMatcher {
-	return ElementMatcher{FieldName: field, FieldValue: value}
+func MatchElement(field string, value string, keys []string, values []string) ElementMatcher {
```
Comment (yuwenma) [inline kyaml/yaml/fns.go:275]: MatchElement is exportable functions. Is that okay to change the passing reference? Will it break its potential callers?

File: kyaml/yaml/fns.go [hunk 5]
```diff
@@ -273,19 +343,28 @@ func (e ElementMatcher) Filter(rn *RNode) (*RNode, error) {
...
-		if IsFoundOrError(field, err) {
-			return elem, err
+			found := true
+			for i, key := range e.Keys {
+				field, err = elem.Pipe(MatchField(key, e.Values[i]))
+				if !IsFoundOrError(field, err) {
+					found = false
+					break
+				}
+			}
```
Comment (yuwenma) [inline kyaml/yaml/fns.go:357]: I'm not quite familiar with this code snippet. The logic looks a little bit awkward. We normally start `found` with false, and break once "found" a match. I guess it may be a naming issue. how about changing "found" to "isMissing". So as the code can be read as "If one key is not found, isMissing is true."

File: kyaml/yaml/rnode.go [hunk 6]
```diff
@@ -474,13 +474,32 @@ func (rn *RNode) ElementValues(key string) ([]string, error) {
...
+	if err := ErrorIfInvalid(rn, yaml.SequenceNode); err != nil {
+		return nil, errors.Wrap(err)
+	}
+	elements := make([][]string, len(rn.Content()))
+
+	for i := 0; i < len(rn.Content()); i++ {
+		for _, key := range keys {
+			field := NewRNode(rn.Content()[i]).Field(key)
+			if !field.IsNilOrEmpty() {
```
Comment (yuwenma) [inline kyaml/yaml/rnode.go:486]: super nit: suggest letting the if statement match its content when possible. 

if field.IsNilOrEmpty() {
    // Attach nil or empty
} else {
    // Attach specific value.
}

File: kyaml/yaml/rnode.go [hunk 7]
```diff
@@ -474,13 +474,32 @@ func (rn *RNode) ElementValues(key string) ([]string, error) {
...
+			}
+		}
+	}
+	return elements, nil
+}
+
 // Element returns the element in the list which contains the field matching the value.
 // Returns nil for non-SequenceNodes or if no Element matches.
-func (rn *RNode) Element(key, value string) *RNode {
+func (rn *RNode) Element(keys []string, values []string) *RNode {
```
Comment (yuwenma) [inline kyaml/yaml/rnode.go:498]: same question: will this passing arg change break existing caller functions?

File: kyaml/yaml/fns.go (hunk 4 above)
Comment (natasha41575) [reply kyaml/yaml/fns.go:275]: The only alternative that I can think of would be to keep all of the exportable functions as is and add new ones that do the same thing except take []string instead of string. Should I try to do that instead?

File: kyaml/yaml/fns.go (hunk 2 above)
Comment (Shell32-Natsu) [inline kyaml/yaml/fns.go:75]: Is it necessary to make this function public?

File: kyaml/yaml/fns.go [hunk 8]
```diff
@@ -124,6 +128,69 @@ func (e ElementSetter) Filter(rn *RNode) (*RNode, error) {
...
+
+	if len(e.Keys) == 0 {
+		if e.Key == "$patch" {
+			return e.FilterPatch(rn)
+		} else {
+			return nil, nil
+		}
+	}
+
+	// build the new Content slice
```
Comment (Shell32-Natsu) [inline kyaml/yaml/fns.go:144]: Any difference compare to line 80?

File: kyaml/yaml/fns.go (hunk 4 above)
Comment (Shell32-Natsu) [reply kyaml/yaml/fns.go:275]: I'd like to suggest to add new methods with more parameters.

File: kyaml/yaml/fns.go [hunk 9]
```diff
@@ -273,19 +343,28 @@ func (e ElementMatcher) Filter(rn *RNode) (*RNode, error) {
...
 			field, err = elem.Pipe(Get(e.FieldName))
+			if IsFoundOrError(field, err) {
+				return elem, err
+			}
 		} else {
-			field, err = elem.Pipe(MatchField(e.FieldName, e.FieldValue))
-		}
-		if IsFoundOrError(field, err) {
-			return elem, err
+			found := true
```
Comment (Shell32-Natsu) [inline kyaml/yaml/fns.go:350]: Please add a comment to explain this.

File: kyaml/yaml/rnode.go (hunk 7 above)
Comment (Shell32-Natsu) [reply kyaml/yaml/rnode.go:498]: `RNode` is one of the most fundamental struct in kyaml. We need to have tons of changes to apply this modification. And I believe this change will break kpt somehow.

**Review by Shell32-Natsu — COMMENTED** [review]
Please add a little bit more comments.

File: kyaml/yaml/fns.go (hunk 4 above)
Comment (natasha41575) [reply kyaml/yaml/fns.go:275]: Thanks Donny, I'm working on making that change now

File: kyaml/openapi/openapi.go [hunk 10]
```diff
@@ -372,13 +403,20 @@ const (
...
 	kubernetesMergeKeyExtensionKey = "x-kubernetes-patch-merge-key"
+
 	// kubernetesPatchStrategyExtensionKey is the key to lookup the kubernetes patch strategy
 	// extension -- the extension is a string
 	kubernetesPatchStrategyExtensionKey = "x-kubernetes-patch-strategy"
 
+
+	// kubernetesMergeKeyMapList is the list of merge keys when there needs to be multiple
+	// -- the extension is an array of strings
+	kubernetesMergeKeyMapList = "x-kubernetes-list-map-keys"
```
Comment (phanimarupaka) [inline kyaml/openapi/openapi.go:418]: Who maintains/updates this extension in the openapi swagger.json ? And how ?

File: kyaml/yaml/walk/associative_sequence.go [hunk 11]
```diff
@@ -125,26 +163,36 @@ func (l *Walker) walkAssociativeSequence() (*yaml.RNode, error) {
...
 	}
 
-	// get the merge key from schema
-	var key, strategy string
+	// get the merge key(s) from schema
+	var strategy string
+	var keys []string
 	if l.Schema != nil {
-		strategy, key = l.Schema.PatchStrategyAndKey()
+		strategy, keys = l.Schema.PatchStrategyAndKeyList()
```
Comment (phanimarupaka) [inline kyaml/yaml/walk/associative_sequence.go:170]: At this point, if openAPI returns multiple keys, why can't we just decide upon the merge key here itself so that the remaining logic can remain unchanged. 

In simple words, we used to get 1(or 0) merge key from openAPI schema. Now we might get multiple. If we get multiple of those, since we have access to l.Sources here, why can't we just pick the merge key here itself?

**Comment by phanimarupaka** [comment]
@natasha41575 Please add tests to [Merge3](https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/yaml/merge3/merge3_test.go) logic as well. This change affects both kpt and kustomize and we need tests to cover kpt(merge3) cases.

File: kyaml/yaml/walk/associative_sequence.go (hunk 11 above)
Comment (natasha41575) [reply kyaml/yaml/walk/associative_sequence.go:170]: If I'm understanding your question correctly, the goal is not to choose a single merge key; it's to use all the merge keys from the schema as a tuple - e.g. for containerPort ports it would be the {containerPort, protocol} tuple, and the test cases I added as part of its PR document the desired behavior

File: kyaml/openapi/openapi.go (hunk 10 above)
Comment (monopole) [reply kyaml/openapi/openapi.go:417]: That's the job of the apimachinery sig.

Using it here is the recommendation of @apelisse from #3125, a pr that was closed as draft.

We should probably come to a conclusion here, and not start a new pr.

File: api/filters/patchstrategicmerge/patchstrategicmerge_test.go [hunk 12]
```diff
@@ -370,6 +370,213 @@ spec:
...
+  name: test-deployment
+spec:
+  template:
+    spec:
+      containers:
+      - image: test-image
+        name: test-deployment
+        ports:
+        - containerPort: 80
+        - containerPort: 8080
```
Comment (monopole) [inline api/filters/patchstrategicmerge/patchstrategicmerge_test.go:579]: thanks, i think this covers the ask from #3125 

none of the `expected` results violates what i'd expect from reading the patches.

To help understand this important change,

please add these tests to the existing code base right now in a separate PR,

but (in that PR) change the expected result to whatever actually happens.

Let's merge that first.

This this PR will be shorter and easier to understand as a behavior change.


To drive this point home, you could add these tests to the existing code base right now,
actually i'm ask

File: kyaml/sets/sets.go [hunk 13]
```diff
@@ -3,6 +3,37 @@
 
 package sets
 
+type StringList [][]string
```
Comment (monopole) [inline kyaml/sets/sets.go:6]: For this to be in this package, it has to be a set of string lists.

But it doesn't act like a set, e.g. one can append the same list twice and it will appear twice.

If that's what we're shooting for, and its important to the rest of the PR, 
please
 - add comment  above the `type` definition showing some examples (i was confused
   by the name and had to read the code, which implies we need a comment).
 - add a test file (sets_test.go) and a test which tests ithat inserting the same thing twice
   doesn't change the size of the set
 - put that in a different PR to get the size of this PR down.

File: kyaml/yaml/fns.go [hunk 14]
```diff
@@ -124,6 +124,100 @@ func (e ElementSetter) Filter(rn *RNode) (*RNode, error) {
 	return NewRNode(e.Element), nil
 }
 
+// ElementSetterList sets the value for an Element in an associative list.  ElementSetterList
```
Comment (monopole) [inline kyaml/yaml/fns.go:127]: Please try to keep line length under 80 chars using linebreaks.
The first sentence here is perfect on a line by itself

Comment (monopole) [reply kyaml/yaml/fns.go:127]: Due to the similarity, the comment here should describe the difference between
this and the ElementSetter struct on line 46 (old file)

Comment (monopole) [reply kyaml/yaml/fns.go:127]: Also, fns_test.go needs to have a couple of regression tests for it.
If we are defining these things as public structs, we should cover them.

Maybe @Shell32-Natsu (last toucher of ElementSetter)
could first add some tests there for ElementSetter (in a different Pr) to establish a pattern.

Nothing fancy, just use the filter in a couple of cases.

Then this PR could follow that pattern.

Comment (monopole) [reply kyaml/yaml/fns.go:127]: Likewise, again, to get the size of this PR down, the ElemenetSetterList and it's tests could be in a PR (or commit) by itself.  I prefer PR.

File: kyaml/yaml/walk/associative_sequence.go (hunk 11 above)
Comment (monopole) [reply kyaml/yaml/walk/associative_sequence.go:259]: right, the primary key for a merge can now be a list of fields instead of just one.

we should aspire to treat the old behavior as the degenerate case of a list of length one.

File: api/filters/patchstrategicmerge/patchstrategicmerge_test.go (hunk 12 above)
Comment (natasha41575) [reply api/filters/patchstrategicmerge/patchstrategicmerge_test.go:579]: thank you! I opened #3170

File: kyaml/sets/sets.go (hunk 13 above)
Comment (natasha41575) [reply kyaml/sets/sets.go:6]: opened #3173, thanks!

File: kyaml/openapi/openapi.go (hunk 10 above)
Comment (apelisse) [reply kyaml/openapi/openapi.go:417]: >> Who maintains/updates this extension in the openapi swagger.json ? And how ?
> That's the job of the apimachinery sig.

Partly. We maintain the machinery, people writing Kubernetes types (from various sigs) and CRDs do use them.

File: kyaml/yaml/fns.go (hunk 14 above)
Comment (natasha41575) [reply kyaml/yaml/fns.go:127]: opened #3176

File: kyaml/sets/sets.go (hunk 13 above)
Comment (monopole) [reply kyaml/sets/sets.go:6]: now merged, thanks

File: kyaml/yaml/walk/associative_sequence.go (hunk 11 above)
Comment (phanimarupaka) [reply kyaml/yaml/walk/associative_sequence.go:259]: @monopole @natasha41575 I am not sure if I agree with this behavior. In this [test case](https://github.com/kubernetes-sigs/kustomize/pull/3159/files#diff-d439562bab858ebaa57ae8c621f357b89626415b2bec9fad2b1ce0443d629c7fR424) for example, why is the expected behavior to append the new section as opposed to update the protocol to `TCP`? As mentioned in the tests we should first get to consensus on what is the expected behavior for `append vs update` before moving further. 

And if merge key can be more than one, why only tuples? Why can't it be combination of 3 keys {containerPort, protocol, name}?

Comment (natasha41575) [reply kyaml/yaml/walk/associative_sequence.go:259]: > And if merge key can be more than one, why only tuples? Why can't it be combination of 3 keys {containerPort, protocol, name}?
You can, if the openAPI schema had {containerPort, protocol, name} in "x-kubernetes-list-map-keys" 
It would also update the protocol rather than append if the openAPI schema had only {containerPort} in "x-kubernetes-list-map-keys" instead of {containerPort, protocol}

Comment (natasha41575) [reply kyaml/yaml/walk/associative_sequence.go:259]: I think what @monopole said in one of our conversations is that with this code, whether or not the behavior is correct is dependent on the keys in the schema now

Comment (apelisse) [reply kyaml/yaml/walk/associative_sequence.go:259]: > whether or not the behavior is correct is dependent on the keys in the schema now

That seems correct.

> And if merge key can be more than one, why only tuples? Why can't it be combination of 3 keys {containerPort, protocol, name}?

The API doesn't use the name as the key because the name is too often missing, this would have broken existing behavior for many people.

> what is the expected behavior for `append vs update`

Jeff said that kustomize typically appends in that case. Removal can be done with a json patch.

**Comment by monopole** [comment]
needs rebase obviously, given all the nice work in other prs.
/ok-to-test

**Comment by monopole** [comment]
Please rebase and squash down to one commit.

**Review by monopole — APPROVED** [review]
yes, 2 commits better given two modules, thanks
